File: framework/pure_nn/optimizers.py
from abc import ABC
from .utils import *


def supported_optimizers():
    return [x.__name__ for x in Optimizer.__subclasses__()]


class Optimizer:
    def __init__(self, params, lr):
        self.params = params
        self.lr = lr

    def apply(self):
        raise NotImplementedError


class SGD(Optimizer, ABC):
    def __init__(self, params, lr):
        super(SGD, self).__init__(params, lr)

    def apply(self):
        for param in self.params.values():
            grad_step_w = rescale(param["dW"], -self.lr)
            param["W"] = mat_add(param["W"], grad_step_w)
            grad_step_b = rescale(param["db"], -self.lr)
            param["b"] = mat_add(param["b"], grad_step_b)


class Momentum(Optimizer, ABC):
    def __init__(self, params, lr, mu):
        super(Momentum, self).__init__(params, lr)
        self.mu = mu
        for layer in list(self.params.values()):
            layer.update({"gW": rescale(layer["dW"], 0)})
            layer.update({"gb": rescale(layer["db"], 0)})

    def apply(self):
        for param in self.params.values():
            param["gW"] = mat_add(param["dW"], rescale(param["gW"], self.mu))
            grad_step_w = rescale(param["gW"], -self.lr)
            param["W"] = mat_add(param["W"], grad_step_w)
            param["gb"] = mat_add(param["db"], rescale(param["gb"], self.mu))
            grad_step_b = rescale(param["gb"], -self.lr)
            param["b"] = mat_add(param["b"], grad_step_b)


class RMSProp(Optimizer, ABC):
    def __init__(self, params, lr=0.01, beta=0.99, eps=1e-8):
        super(RMSProp, self).__init__(params, lr)
        self.beta = beta
        self.eps = eps
        for layer in list(self.params.values()):
            layer.update({"sW": rescale(layer["dW"], 0)})
            layer.update({"sb": rescale(layer["db"], 0)})

    def apply(self):
        for param in self.params.values():

            grad_square_w = element_wise_mul(param["dW"], param["dW"])
            grad_square_w = rescale(grad_square_w, 1 - self.beta)
            param["sW"] = mat_add(rescale(param["sW"], self.beta), grad_square_w)
            grad_step_w = element_wise_mul(param["dW"], element_wise_rev(add_scalar(mat_sqrt(param["sW"]), self.eps)))
            param["W"] = mat_add(param["W"], rescale(grad_step_w, -self.lr))

            grad_square_b = element_wise_mul(param["db"], param["db"])
            grad_square_b = rescale(grad_square_b, 1 - self.beta)
            param["sb"] = mat_add(rescale(param["sb"], self.beta), grad_square_b)
            grad_step_b = element_wise_mul(param["db"], element_wise_rev(add_scalar(mat_sqrt(param["sb"]), self.eps)))
            param["b"] = mat_add(param["b"], rescale(grad_step_b, -self.lr))


class AdaGrad(Optimizer, ABC):
    def __init__(self, params, lr=0.01, eps=1e-8):
        super(AdaGrad, self).__init__(params, lr)
        self.eps = eps
        for layer in list(self.params.values()):
            layer.update({"sW": rescale(layer["dW"], 0)})
            layer.update({"sb": rescale(layer["db"], 0)})

    def apply(self):
        for param in self.params.values():

            grad_square_w = element_wise_mul(param["dW"], param["dW"])
            param["sW"] = mat_add(param["sW"], grad_square_w)
            grad_step_w = element_wise_mul(param["dW"], element_wise_rev(add_scalar(mat_sqrt(param["sW"]), self.eps)))
            param["W"] = mat_add(param["W"], rescale(grad_step_w, -self.lr))

            grad_square_b = element_wise_mul(param["db"], param["db"])
            param["sb"] = mat_add(param["sb"], grad_square_b)
            grad_step_b = element_wise_mul(param["db"], element_wise_rev(add_scalar(mat_sqrt(param["sb"]), self.eps)))
            param["b"] = mat_add(param["b"], rescale(grad_step_b, -self.lr))


class Adam(Optimizer, ABC):
    def __init__(self, params, lr=0.001, beta1=0.9, beta2=0.999, eps=1e-8):
        super(Adam, self).__init__(params, lr)
        self.beta1 = beta1
        self.beta2 = beta2
        self.eps = eps
        self.k = 1
        for layer in list(self.params.values()):
            layer.update({"mW": rescale(layer["dW"], 0)})
            layer.update({"vW": rescale(layer["dW"], 0)})
            layer.update({"mb": rescale(layer["db"], 0)})
            layer.update({"vb": rescale(layer["db"], 0)})

    def apply(self):
        for param in self.params.values():
            param["mW"] = mat_add(rescale(param["dW"], 1 - self.beta1), rescale(param["mW"], self.beta1))
            param["vW"] = mat_add(rescale(element_wise_mul(param["dW"], param["dW"]), 1 - self.beta2), rescale(param["vW"], self.beta2))
            mw_hat = rescale(param["mW"],  1 / (1 - self.beta1 ** self.k))
            vw_hat = rescale(param["vW"], 1 / (1 - self.beta2 ** self.k))
            grad_step_w = element_wise_mul(mw_hat, element_wise_rev(add_scalar(mat_sqrt(vw_hat), self.eps)))
            param["W"] = mat_add(param["W"], rescale(grad_step_w, -self.lr))

            param["mb"] = mat_add(rescale(param["db"], 1 - self.beta1), rescale(param["mb"], self.beta1))
            param["vb"] = mat_add(rescale(element_wise_mul(param["db"], param["db"]), 1 - self.beta2), rescale(param["vb"], self.beta2))
            mb_hat = rescale(param["mb"], 1 / (1 - self.beta1 ** self.k))
            vb_hat = rescale(param["vb"], 1 / (1 - self.beta2 ** self.k))
            grad_step_b = element_wise_mul(mb_hat, element_wise_rev(add_scalar(mat_sqrt(vb_hat), self.eps)))
            param["b"] = mat_add(param["b"], rescale(grad_step_b, -self.lr))
        self.k += 1

File: cpp_nn/optimizers.cpp
#include "optimizers.h"

Optimizer::Optimizer(float lr, vector<Layer*> &params)
{
    this->lr = lr;
    this->parameters = params;
}

void SGD::apply()
{
    for(size_t i = 0; i < this->parameters.size(); i++){
        this->parameters[i]->W = this->utils.mat_add(this->parameters[i]->W,
                                                     this->utils.rescale(this->parameters[i]->dW, -this->lr));
        this->parameters[i]->b = this->utils.mat_add(this->parameters[i]->b,
                                                     this->utils.rescale(this->parameters[i]->db, -this->lr));
    }
}

Momentum::Momentum(float lr, float mu, vector<Layer*> &params) : Optimizer{lr, params}{
    this->mu = mu;
    for(size_t i = 0; i < this->parameters.size(); i++){
        this->gW.push_back(this->utils.rescale(this->parameters[i]->W, 0.0));
        this->gb.push_back(this->utils.rescale(this->parameters[i]->b, 0.0));
    }
}

void Momentum::apply()
{
    for (size_t i = 0; i < this->parameters.size(); i++){
        this->gW[i] = this->utils.mat_add(this->parameters[i]->dW,
                                          this->utils.rescale(this->gW[i], this->mu)
                                          );
        this->parameters[i]->W = this->utils.mat_add(this->parameters[i]->W,
                                                     this->utils.rescale(this->gW[i], -this->lr)
                                                     );
        this->gb[i] = this->utils.mat_add(this->parameters[i]->db,
                                          this->utils.rescale(this->gb[i], this->mu)
                                          );
        this->parameters[i]->b = this->utils.mat_add(this->parameters[i]->b,
                                                     this->utils.rescale(this->gb[i], -this->lr)
                                                     );
    }

}

RMSProp::RMSProp(float lr, float beta, vector<Layer*> &params) : Optimizer(lr, params){
    this->beta = beta;
    for (size_t i = 0; i < this->parameters.size(); i++){
        this->sW.push_back(this->utils.rescale(this->parameters[i]->W, 0.0));
        this->sb.push_back(this->utils.rescale(this->parameters[i]->b, 0.0));

    }

}

void RMSProp::apply()
{
    for (size_t i = 0; i < this->parameters.size(); i++){
        float_batch grad_square_w = this->utils.element_wise_mul(this->parameters[i]->dW, this->parameters[i]->dW);
        grad_square_w = this->utils.rescale(grad_square_w, 1 - this->beta);
        this->sW[i] = this->utils.mat_add(this->utils.rescale(this->sW[i], beta), grad_square_w);
        float_batch grad_step_w = this->utils.element_wise_mul(this->parameters[i]->dW,
                                                               this->utils.element_wise_rev(
                                                                   this->utils.add_scalar(
                                                                       this->utils.element_wise_sqrt(this->sW[i]), eps)
                                                                   )
                                                               );
        this->parameters[i]->W = this->utils.mat_add(this->parameters[i]->W, this->utils.rescale(grad_step_w, -this->lr));

        float_batch grad_square_b = this->utils.element_wise_mul(this->parameters[i]->db, this->parameters[i]->db);
        grad_square_b = this->utils.rescale(grad_square_b, 1 - this->beta);
        this->sb[i] = this->utils.mat_add(this->utils.rescale(this->sb[i], beta), grad_square_b);
        float_batch grad_step_b = this->utils.element_wise_mul(this->parameters[i]->db,
                                                               this->utils.element_wise_rev(
                                                                   this->utils.add_scalar(
                                                                       this->utils.element_wise_sqrt(this->sb[i]), eps)
                                                                   )
                                                               );
        this->parameters[i]->b = this->utils.mat_add(this->parameters[i]->b, this->utils.rescale(grad_step_b, -this->lr));
    }
}

AdaGrad::AdaGrad(float lr, vector<Layer*> &params) : Optimizer(lr, params)
{
    for (size_t i = 0; i < this->parameters.size(); i++){
        this->sW.push_back(this->utils.rescale(this->parameters[i]->W, 0.0));
        this->sb.push_back(this->utils.rescale(this->parameters[i]->b, 0.0));

    }
}

void AdaGrad::apply()
{
    for (size_t i = 0; i < this->parameters.size(); i++){
        float_batch grad_square_w = this->utils.element_wise_mul(this->parameters[i]->dW, this->parameters[i]->dW);
        this->sW[i] = this->utils.mat_add(this->sW[i], grad_square_w);
        float_batch grad_step_w = this->utils.element_wise_mul(this->parameters[i]->dW,
                                                               this->utils.element_wise_rev(
                                                                   this->utils.add_scalar(
                                                                       this->utils.element_wise_sqrt(this->sW[i]), eps)
                                                                   )
                                                               );
        this->parameters[i]->W = this->utils.mat_add(this->parameters[i]->W, this->utils.rescale(grad_step_w, -this->lr));

        float_batch grad_square_b = this->utils.element_wise_mul(this->parameters[i]->db, this->parameters[i]->db);
        this->sb[i] = this->utils.mat_add(this->sb[i], grad_square_b);
        float_batch grad_step_b = this->utils.element_wise_mul(this->parameters[i]->db,
                                                               this->utils.element_wise_rev(
                                                                   this->utils.add_scalar(
                                                                       this->utils.element_wise_sqrt(this->sb[i]), eps)
                                                                   )
                                                               );
        this->parameters[i]->b = this->utils.mat_add(this->parameters[i]->b, this->utils.rescale(grad_step_b, -this->lr));


    }
}

Adam::Adam(float lr, float beta1, float beta2, vector<Layer*> &params) : Optimizer(lr, params)
{
    this->beta1 = beta1;
    this->beta2 = beta2;
    for (size_t i = 0; i < this->parameters.size(); i++){
        this->mW.push_back(this->utils.rescale(this->parameters[i]->W, 0.0));
        this->vW.push_back(this->utils.rescale(this->parameters[i]->W, 0.0));
        this->mb.push_back(this->utils.rescale(this->parameters[i]->b, 0.0));
        this->vb.push_back(this->utils.rescale(this->parameters[i]->b, 0.0));

    }
}

void Adam::apply()
{
    for (size_t i = 0; i < this->parameters.size(); i++){
        this->mW[i] = this->utils.mat_add(this->utils.rescale(this->parameters[i]->dW, 1 - this->beta1),
                                          this->utils.rescale(this->mW[i], this->beta1)
                                          );
        this->vW[i] = this->utils.mat_add(this->utils.rescale(this->utils.element_wise_mul(this->parameters[i]->dW, this->parameters[i]->dW), 1 - this->beta2),
                                          this->utils.rescale(this->vW[i], this->beta2)
                                          );
        float_batch mW_hat = this->utils.rescale(this->mW[i], 1 / (1 - pow(this->beta1, this->k)));
        float_batch vW_hat = this->utils.rescale(this->vW[i], 1 / (1 - pow(this->beta2, this->k)));
        float_batch grad_step_w = this->utils.element_wise_mul(mW_hat,
                                                               this->utils.element_wise_rev(
                                                               this->utils.add_scalar(
                                                               this->utils.element_wise_sqrt(vW_hat), eps)
                                                                   )
                                                               );
        this->parameters[i]->W = this->utils.mat_add(this->parameters[i]->W, this->utils.rescale(grad_step_w, -this->lr));
        //---------------//
        this->mb[i] = this->utils.mat_add(this->utils.rescale(this->parameters[i]->db, 1 - this->beta1),
                                          this->utils.rescale(this->mb[i], this->beta1)
                                          );
        this->vb[i] = this->utils.mat_add(this->utils.rescale(this->utils.element_wise_mul(this->parameters[i]->db, this->parameters[i]->db), 1 - this->beta2),
                                          this->utils.rescale(this->vb[i], this->beta2)
                                          );
        float_batch mb_hat = this->utils.rescale(this->mb[i], 1 / (1 - pow(this->beta1, this->k)));
        float_batch vb_hat = this->utils.rescale(this->vb[i], 1 / (1 - pow(this->beta2, this->k)));
        float_batch grad_step_b = this->utils.element_wise_mul(mb_hat,
                                                               this->utils.element_wise_rev(
                                                               this->utils.add_scalar(
                                                               this->utils.element_wise_sqrt(vb_hat), eps)
                                                                   )
                                                               );
        this->parameters[i]->b = this->utils.mat_add(this->parameters[i]->b, this->utils.rescale(grad_step_b, -this->lr));

    }
    this->k++;
}
